### PR TITLE
first pass at positioning jwt vc as an sd-jwt vc without selectively disclosable claims

### PR DESCRIPTION
**Pull request by Sakurann**
alternative to #120 based on @bc-pi 's comment "I can't help but wonder if it'd make sense to have the spec be only about SD-JWT VCs and treat the plain JWT VC as just a special case of an SD-JWT with no selectively disclosable claims."

no changes to the basic structure. simple clarifications what to do when there are no selectively disclosable claims ie no `_sd` claim and no Disclosures. Might need some more details but this is how it might look like.

(cc @OR13, @mprorock)

[linked to rendered preview](https://vcstuff.github.io/draft-terbu-sd-jwt-vc/jwt-vc-2/draft-terbu-sd-jwt-vc.html
)

### CODE REVIEW

File: draft-terbu-sd-jwt-vc.md
Comment (tlodderstedt) [inline draft-terbu-sd-jwt-vc.md:34]: why do we even need to change this?

File: draft-terbu-sd-jwt-vc.md
Comment (tlodderstedt) [inline draft-terbu-sd-jwt-vc.md:102]:
```suggestion
their semantics. SD-JWT can also be used when there are no selectively disclosable claims.
```

File: draft-terbu-sd-jwt-vc.md
Comment (tlodderstedt) [inline draft-terbu-sd-jwt-vc.md:104]:
```suggestion
can be expressed using SD-JWT with and without selective disclosure.
```

File: draft-terbu-sd-jwt-vc.md
Comment (tlodderstedt) [inline draft-terbu-sd-jwt-vc.md:262]: out of curiosity - the change to align issuance and presentation format would lead to the fact that non sd VCs would be JWTs + a `~`, right?

File: draft-terbu-sd-jwt-vc.md
Comment (tlodderstedt) [inline draft-terbu-sd-jwt-vc.md:101]: suggest to reorganize this paragraph 
Move 2nd and 3rd sentence to previous paragraph forward (all related to sd and non sd). Modify 3rd sentence.

"This specification therefore defines how Verifiable Credentials
can be expressed using SD-JWT covering both options, with and without selective disclosure."

1st sentence fits better with the last paragraph as this is about the content. Suggest to add the following sentence between this sentence and the rest: 
"SD-JWT VC defines a content model based on JWT claims."

**Review by tlodderstedt — CHANGES_REQUESTED** [review]

File: draft-terbu-sd-jwt-vc.md
Comment (paulbastian) [reply draft-terbu-sd-jwt-vc.md:34]: while this might be technically correct, I agree it might be confusing to first time readers. Either reverse to:
"based on the SD-JWT (Selective Disclosure for JWT and JWS with JSON payloads) format"
or leave it as is, as the full title is gehind the reference

File: draft-terbu-sd-jwt-vc.md
Comment (paulbastian) [inline draft-terbu-sd-jwt-vc.md:262]:
```suggestion
In that case, the SD-JWT VC MUST NOT contain the `_sd` claim in the JWT body and
MUST NOT have any Disclosures.
```

Comment (paulbastian) [reply draft-terbu-sd-jwt-vc.md:262]: I guess so, also key binding would be very easy as it is still being appended after ~ in presentation?

File: draft-terbu-sd-jwt-vc.md
Comment (paulbastian) [inline draft-terbu-sd-jwt-vc.md:300]:
```suggestion
If there are no selectively disclosable claims, there is no need to process the
`_sd` claim nor any Disclosures.
```

File: draft-terbu-sd-jwt-vc.md
Comment (paulbastian) [inline draft-terbu-sd-jwt-vc.md:438]:
```suggestion
When there are no selectively disclosable claims, a presentation of SD-JWT VC
does not contain any Disclosures.
```

File: draft-terbu-sd-jwt-vc.md
Comment (paulbastian) [inline draft-terbu-sd-jwt-vc.md:96]: I actually like this section as it gives rationale for the issuer-holder-verifier model

File: draft-terbu-sd-jwt-vc.md
Comment (paulbastian) [reply draft-terbu-sd-jwt-vc.md:101]: I agree, when the section above is removed, the paragraphs don't really fit together nicely

**Comment by paulbastian** [comment]
Generally, I really like this approach!
Some examples could be useful for key binding with JWT (no disclosures)

**Comment by OR13** [comment]
I'd expect the typ for both to have +jwt if that were really the case.

There is another draft up for multiple suffixes.

https://www.ietf.org/archive/id/draft-ietf-mediaman-suffixes-04.html

Is it simple enough to support now?

File: draft-terbu-sd-jwt-vc.md
Comment (tlodderstedt) [reply draft-terbu-sd-jwt-vc.md:34]:
```suggestion
rules to express Verifiable Credentials with JSON payloads based on the Selective Disclosure
```

File: draft-terbu-sd-jwt-vc.md
Comment (tlodderstedt) [inline draft-terbu-sd-jwt-vc.md:35]:
```suggestion
for JWT (SD-JWT) format [@!I-D.ietf-oauth-selective-disclosure-jwt].
```

File: draft-terbu-sd-jwt-vc.md
Comment (tlodderstedt) [inline draft-terbu-sd-jwt-vc.md:36]:
```suggestion
```

File: draft-terbu-sd-jwt-vc.md
Comment (tlodderstedt) [reply draft-terbu-sd-jwt-vc.md:96]: I think this section distracts from the real topic.

File: draft-terbu-sd-jwt-vc.md
Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md:34]: because I think acronyms should be expanded when used the first time. we are deep in the details, but many do not know what SD-JWT is when they see it for the first time. and we should not assume that anyone reading this already knows what SD-JWT is.

File: draft-terbu-sd-jwt-vc.md
Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md:260]:
```suggestion
In that case, the SD-JWT VC MUST NOT contain the `_sd` claim in the JWT body. I also
MUST NOT have any Disclosures.
```

Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md:260]: I prefer avoiding two normative requirements (MUST NOTs) in the same sentence.

File: draft-terbu-sd-jwt-vc.md
Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md]: thanks - I think it reads better now.

File: draft-terbu-sd-jwt-vc.md
Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md]: outdated with torsten's commit.

File: draft-terbu-sd-jwt-vc.md
Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md]: outdated with torsten's commit.

File: draft-terbu-sd-jwt-vc.md
Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md]: aaa that is an important question. It is a little counter intuitive, but if sd-jwt is the basis of jwt vcs ending with Tilde will be more natural and will be a nice differentiator between normal jwts and jwt vcs..?

File: draft-terbu-sd-jwt-vc.md
Comment (bc-pi) [inline draft-terbu-sd-jwt-vc.md:259]:
```suggestion
In that case, the SD-JWT VC MUST NOT contain the `_sd` claim in the JWT body. It also
```

**Comment by mprorock** [comment]
should this be titled `draft-terbu-oauth-vc-jwt` ?

**Comment by Sakurann** [comment]
Yes. Will make a change once we have a decision whether PR #120 or this one is the way to go…

**Review by OR13 — APPROVED** [review]

File: draft-terbu-sd-jwt-vc.md
Comment (alenhorvat) [reply draft-terbu-sd-jwt-vc.md:34]: I guess this should be aligned with: https://github.com/vcstuff/draft-terbu-sd-jwt-vc/pull/132

**Comment by alenhorvat** [comment]
What is the position of this specification?
From my understanding
- it defines a VC with JWT claims
- it defines a profile for the JWT claims
- it defines a profile for securing the VC

There's an open discussion between the sd-jwt and JWS protection. Relevant discussion has been started https://github.com/oauth-wg/oauth-selective-disclosure-jwt/issues/291 - sd-jwt introduces an encoding that becomes incompatible with JWS in the limit where selective disclosure is not used. In other words, it introduces a first-order phase transition. Let's see how this topic resolves.

IMO, the data model and securing mechanisms should support both sd-jwt and JWS.

File: draft-terbu-sd-jwt-vc.md
Comment (alenhorvat) [reply draft-terbu-sd-jwt-vc.md:36]: I agree that the supported securing mechanisms should be explicit.

Should the specification define which claims can be hidden and which not when SD is applied? e.g., iss, exp, nbf, cnf is probably conditional.

Comment (paulbastian) [reply draft-terbu-sd-jwt-vc.md:36]: This is already the case.
> If present, the following registered JWT claims MUST be included in the SD-JWT and MUST NOT be included in the Disclosures, i.e. cannot be selectively disclosed:

File: draft-terbu-sd-jwt-vc.md
Comment (danielfett) [inline draft-terbu-sd-jwt-vc.md:258]:
```suggestion
SD-JWT VC MAY have no selectively disclosable claims.
```

File: draft-terbu-sd-jwt-vc.md
Comment (danielfett) [inline draft-terbu-sd-jwt-vc.md:578]:
```suggestion
We would like to thank
Alen Horvat,
Andres Uribe,
Brian Campbell,
Christian Bormann,
Giuseppe De Marco,
Michael Jones,
Mike Prorock,
Orie Steele,
Paul Bastian,
Torsten Lodderstedt,
Tobias Looker, and
Kristina Yasuda
for their contributions (some of which substantial) to this draft and to the initial set of implementations.
```

File: draft-terbu-sd-jwt-vc.md
Comment (danielfett) [inline draft-terbu-sd-jwt-vc.md:260]:
```suggestion
MUST NOT have any Disclosures. For issuance, the trailing `~` MUST BE omitted from the SD-JWT.
For presentation the trailing `~` MUST BE omitted unless a Key Binding JWT is present.
```

**Comment by danielfett** [comment]
I do like this approach, but I am a bit concerned that even in the 'no disclosures' case, we'll have the trailing tilde in the issuance and presentation (when no KB is present). A JWT is *not* a special case of an SD-JWT for this reason.

**Comment by alenhorvat** [comment]
> I do like this approach, but I am a bit concerned that even in the 'no disclosures' case, we'll have the trailing tilde in the issuance and presentation (when no KB is present). A JWT is _not_ a special case of an SD-JWT for this reason.

I agree with the concern.

**Comment by bc-pi** [comment]
> I do like this approach, but I am a bit concerned that even in the 'no disclosures' case, we'll have the trailing tilde in the issuance and presentation (when no KB is present). A JWT is _not_ a special case of an SD-JWT for this reason.

IMHO keeping the trailing tilde is preferable as it keeps the model consistent.

File: draft-terbu-sd-jwt-vc.md
Comment (selfissued) [reply draft-terbu-sd-jwt-vc.md:35]: I agree that the "with JSON payloads" wording is clunky and should go.  All JWTs have JSON payloads, so we don't need to say so in the title.

Comment (selfissued) [reply draft-terbu-sd-jwt-vc.md:35]: Also, see https://github.com/vcstuff/draft-terbu-sd-jwt-vc/pull/132, which requests a parallel change to the title.

**Review by selfissued — APPROVED** [review]
I prefer https://github.com/vcstuff/draft-terbu-sd-jwt-vc/pull/120 since I find its description of what we're achieving more natural / less surprising.  But I can live with this description as well, provided the distracting "with JSON Claims" language is removed.

File: draft-terbu-sd-jwt-vc.md
Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md:35]: #132 should be addressed after we have clarity on direction whether #120 or #131. 
jfyi "Selective Disclosure for JWT and JWS with JSON payloads (SD-JWT)" is a new spec title for SD-JWT, and "JWS with JSON payloads" is meant to cater for JSON serialization which is not allowed in JWT>

File: draft-terbu-sd-jwt-vc.md
Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md:34]: let's bikeshed the title in this PR for what is in scope for this PR. and then see if it is aligned or not with #132.

File: draft-terbu-sd-jwt-vc.md
Comment (bc-pi) [reply draft-terbu-sd-jwt-vc.md:35]: The full (and admittedly clunky) title of SD-JWT doesn't need to be spelled out here though.  (and ugh we should have capitalized Payloads in that title)  
```suggestion
for JWT (SD-JWT) [@!I-D.ietf-oauth-selective-disclosure-jwt] format.
```

File: draft-terbu-sd-jwt-vc.md
Comment (alenhorvat) [reply draft-terbu-sd-jwt-vc.md:34]: I believe it is important to clarify what the specification is trying to achieve.
JWS: compact or JSON serialised JWS -> protection mechanism + serialisation
SD-JWT: SD-JWT serialisation (JWS+disclosures+tilde) or JSON serialised JWS -> protection mechanism + serialisation --> note that this could be fixed if the SD authors agree to. Issue is still open, a viable proposal is on the table.
JWT: compact serialised JWS -> claim set + protection
id token -> JWT + rules for the claims -> claim set + verification rules + protection

If I understand correctly, this specification defines a profile for Verifiable Credentials with the JWT claims set. So
vc+jwt -> claim set (VC + JWT claims), verification rules + protection (JWS or SD-JWT)

If my understanding is correct and if the specification is defining a new type, like in #132: vc+jwt, the specification can be defined as:
Verifiable Credential with JWT claims and protected by JWS (any serialisation) and/or SD-JWT

Or define a new brand:
- Verifiable Web Tokens (I prefer this one)
- JSON Verifiable Credentials
- JSON Web Credentials
- Verifiable Credential Tokens (don't like this one)

I agree with @selfissued  that "with JSON" payloads is clumsy/confusing.

Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md:34]: I don't think things to be that complicated...

Bottomline, there is a need for a credential format that uses JWT claim set in the payload, and depending on the use-case there may or may not be a need for selective disclosure. To achieve that we are exploring two approaches:
1. define a credential format without SD based on JWT RFC (JWT VC) and with SD based on the IETF SD-JWT draft (SD-JWT VC) - PR #120
2. define a credential format based on IETF SD-JWT that can be with or without SD (SD-WT VC) - PR #131

that's it.

Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md:34]: from what I have observed, majority seems to be favorable towards the approach in PR #131, because it gives one credential format and not two (consolidation, yay!), with one major discussion point being a `~` after a JWT VC when it is issued. 

a bit of the context... this trailing `~` is the result of the decision being taken in SD-JWT spec to better align the issuance and presentation formats, ([this PR](https://github.com/oauth-wg/oauth-selective-disclosure-jwt/pull/293) does it, and it has not been merged yet). One option is to keep issuance and presentation formats separate in SD-JWT (as-is) so that there is no `~` when (SD-)JWT VC is issued. but that is not that clean from a pure SD-JWT processing perspective is I think the opinion of its authors (@danielfett, @bc-pi). 

so the question is really what to do with a trailing `~` when there is no SD in SD-JWT based VC.

**Comment by Sakurann** [comment]
from what I am sensing, there seems to be a lot of preference towards the approach in this PR over #120, partially because it is simpler and gives one credential format and not two (consolidation, yay!), with one major discussion point being a ~ after a JWT VC when it is issued.

a bit of the context... this trailing ~ is the result of the decision being taken in SD-JWT spec to better align the issuance and presentation formats, (https://github.com/oauth-wg/oauth-selective-disclosure-jwt/pull/293 does it, and it has not been merged yet). One option is to keep issuance and presentation formats separate in SD-JWT (as-is) so that there is no ~ when (SD-)JWT VC is issued. but that is not that clean from a pure SD-JWT processing perspective is I think the opinion of its authors (@danielfett, @bc-pi).

so the question is really what to do with a trailing ~ when there is no SD in SD-JWT based VC.

File: draft-terbu-sd-jwt-vc.md
Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md:260]: if we keep approach in this PR, `~` needs to stay even when there is no SD.

File: draft-terbu-sd-jwt-vc.md
Comment (Sakurann) [inline draft-terbu-sd-jwt-vc.md:35]:
```suggestion
for JWTs (SD-JWT) [@!I-D.ietf-oauth-selective-disclosure-jwt] format.
```

File: draft-terbu-sd-jwt-vc.md
Comment (alenhorvat) [reply draft-terbu-sd-jwt-vc.md:34]: Hi @Sakurann . I agree that id doesn't have to be that complicated. IMO 2nd approach is cleaner as you can clearly split the data model and protection mechanisms.
See:
> If I understand correctly, this specification defines a profile for Verifiable Credentials with the JWT claims set. So
> vc+jwt -> claim set (VC + JWT claims), verification rules + protection (JWS or SD-JWT)

From your answer I understand you agree with "his specification defines a profile for Verifiable Credentials with the JWT claims set"
From design perspective, proposal 2 is better. Having a data model that depends on the signature format has limited re-usability. Let's assume a new signature or SD for JWT format is introduced. Would this required new data model specifications?

File: draft-terbu-sd-jwt-vc.md
Comment (tlodderstedt) [reply draft-terbu-sd-jwt-vc.md:36]: I don't think it needs to be highlighted in the introduction. Rationale is sufficient in my opinion.

**Review by paulbastian — APPROVED** [review]

**Review by tplooker — APPROVED** [review]

File: draft-terbu-sd-jwt-vc.md
Comment (Sakurann) [reply draft-terbu-sd-jwt-vc.md:36]: in other issue.

**Review by c2bo — APPROVED** [review]

**Review by paulbastian — APPROVED** [review]
I approve this approach over #120 as,
- the specification is much cleaner
- I expect implementations to be much cleaner
- I don't see confusion with trailing tilde, because typ is not jwt but sd-jwt
- most interop profiles will require SD support anyway

**Review by tlodderstedt — APPROVED** [review]

**Review by danielfett — APPROVED** [review]

**Comment by Sakurann** [comment]
should be good to merge once @awoie takes a look and approves

**Review by awoie — APPROVED** [review]
approving but will create an issue to address trailing tilde.

**Comment by awoie** [comment]
Merging this PR since a lot of approvals and no objections.

**Review by cghost420 — APPROVED** [review]

**Pull request merged**
